### PR TITLE
initial dynamic subscribe/unsubscribe for attestations to/from subnets

### PR DESCRIPTION
**Pull request by tersec**
This should reduce attestation traffic on the nbc side, as well as pressure on the gossip side, with larger numbers of validators.

### CODE REVIEW

**Comment by tersec** [comment]
```
$ N=0; while ./scripts/launch_local_testnet.sh --testnet 0 --nodes 4 --disable-htop -- --verify-finalization --stop-at-epoch=8; do N=$((N+1)); echo "That was run #${N}"; sleep 67; done
Building: build/beacon_node
Building: build/deposit_contract
INF 2020-08-09 19:08:31.941+02:00 Generating deposits                        tid=158221 file=keystore_management.nim:143 totalValidators=128 validatorsDir=local_testnet_data/validators secretsDir=local_testnet_data/secrets
INF 2020-08-09 19:09:08.687+02:00 Deposit data written                       tid=158221 file=deposit_contract.nim:195 filename=local_testnet_data/deposits.json
Wrote local_testnet_data/network_dir/genesis.ssz
Wrote local_testnet_data/network_dir/bootstrap_nodes.txt
Wrote local_testnet_data/network.json:
{
  "runtimePreset": {
    "MIN_GENESIS_ACTIVE_VALIDATOR_COUNT": 128,
    "MIN_GENESIS_TIME": 0,
    "GENESIS_DELAY": 10,
    "GENESIS_FORK_VERSION": "0x00000000"
  },
  "depositContractAddress": "0x0000000000000000000000000000000000000000",
  "depositContractDeployedAt": "0x0000000000000000000000000000000000000000000000000000000000000000"
}
That was run #1
Building: build/beacon_node
Building: build/deposit_contract
INF 2020-08-09 19:19:32.046+02:00 Generating deposits                        tid=161400 file=keystore_management.nim:143 totalValidators=128 validatorsDir=local_testnet_data/validators secretsDir=local_testnet_data/secrets
INF 2020-08-09 19:20:15.531+02:00 Deposit data written                       tid=161400 file=deposit_contract.nim:195 filename=local_testnet_data/deposits.json
Wrote local_testnet_data/network_dir/genesis.ssz
Wrote local_testnet_data/network_dir/bootstrap_nodes.txt
Wrote local_testnet_data/network.json:
{
  "runtimePreset": {
    "MIN_GENESIS_ACTIVE_VALIDATOR_COUNT": 128,
    "MIN_GENESIS_TIME": 0,
    "GENESIS_DELAY": 10,
    "GENESIS_FORK_VERSION": "0x00000000"
  },
  "depositContractAddress": "0x0000000000000000000000000000000000000000",
  "depositContractDeployedAt": "0x0000000000000000000000000000000000000000000000000000000000000000"
}
That was run #2
Building: build/beacon_node
Building: build/deposit_contract
INF 2020-08-09 19:29:40.845+02:00 Generating deposits                        tid=163994 file=keystore_management.nim:143 totalValidators=128 validatorsDir=local_testnet_data/validators secretsDir=local_testnet_data/secrets
INF 2020-08-09 19:30:18.139+02:00 Deposit data written                       tid=163994 file=deposit_contract.nim:195 filename=local_testnet_data/deposits.json
Wrote local_testnet_data/network_dir/genesis.ssz
Wrote local_testnet_data/network_dir/bootstrap_nodes.txt
Wrote local_testnet_data/network.json:
{
  "runtimePreset": {
    "MIN_GENESIS_ACTIVE_VALIDATOR_COUNT": 128,
    "MIN_GENESIS_TIME": 0,
    "GENESIS_DELAY": 10,
    "GENESIS_FORK_VERSION": "0x00000000"
  },
  "depositContractAddress": "0x0000000000000000000000000000000000000000",
  "depositContractDeployedAt": "0x0000000000000000000000000000000000000000000000000000000000000000"
}
That was run #3
```

**Comment by arnetheduck** [comment]
hm, technically, by not subscribing to the attestations, we're mainly saving processing when we would be idle otherwise? ie the necessary attestation processing (because we're validating for example) still has to be performed, and in the same amounts?

**Comment by tersec** [comment]
> hm, technically, by not subscribing to the attestations, we're mainly saving processing when we would be idle otherwise? ie the necessary attestation processing (because we're validating for example) still has to be performed, and in the same amounts?

Distributed over time, yes, for the most part, though we wouldn't have to ever process attestations in subnets for committees for which the node doesn't have validators if they weren't picked up in either aggregated attestations or beacon blocks. Otherwise, it's saving gossip traffic, but that's mostly it, because the attestation pool will quickly recognize duplicate/overlapping attestations and drop them.

**Comment by tersec** [comment]
```
[Suite] PeerPool testing suite
  [OK] addPeerNoWait() test
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/tests/testutil.nim(81) test_peer_pool
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nimbus-build-system/vendor/Nim/lib/pure/unittest.nim(647) runSuite
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nim-chronos/chronos/asyncloop.nim(1033) waitFor
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nim-chronos/chronos/asyncloop.nim(343) poll
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/tests/test_peer_pool.nim(85) testAddPeer1
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(29) failedAssertImpl
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nimbus-build-system/vendor/Nim/lib/system/assertions.nim(22) raiseAssert
/Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/vendor/nimbus-build-system/vendor/Nim/lib/system/fatal.nim(49) sysFatal

    Unhandled exception: /Users/jenkins/workspace/bus_nim-beacon-chain_prs_PR-1462/tests/test_peer_pool.nim(85, 15) `fut2.finished == true and fut2.failed == false`  [AssertionError]
  [FAILED] addPeer() test
```
Which is odd.

File: beacon_chain/beacon_node.nim [hunk 1]
```diff
@@ -388,6 +389,61 @@ func verifyFinalization(node: BeaconNode, slot: Slot) =
...
+  var attestationSubscriptions: seq[Future[void]] = @[]
+
+  # https://github.com/ethereum/eth2.0-specs/blob/v0.12.2/specs/phase0/p2p-interface.md#attestations-and-aggregation
+  for subnet in subnets:
+    attestationSubscriptions.add(node.network.subscribe(
+      getAttestationTopic(node.forkDigest, subnet),
+      attestationHandler,
+    ))
+
+  waitFor allFutures(attestationSubscriptions)
```
Comment (zah) [inline beacon_chain/beacon_node.nim:411]: According to the spec, you must also update the `node.network.metadata` field here:
https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/p2p-interface.md#metadata

Comment (tersec) [reply beacon_chain/beacon_node.nim:411]: https://github.com/status-im/nim-beacon-chain/pull/1462/commits/4a17389ff125dca9e52f97015cf7a6e1df6b0d52

**Pull request merged**